File: T3/codigos2021/T3_Elementos/app/src/main/java/com/borja/t3_elementos/utils/Coche.java
package com.borja.t3_elementos.utils;

public class Coche {

    private String modelo,marca;
    private int cv, km, imagen;

    public Coche(String modelo, String marca, int cv, int km) {
        this.modelo = modelo;
        this.marca = marca;
        this.cv = cv;
        this.km = km;
    }

    public Coche(String modelo, String marca, int cv, int km, int imagen) {
        this.modelo = modelo;
        this.marca = marca;
        this.cv = cv;
        this.km = km;
        this.imagen = imagen;
    }

    public int getImagen() {
        return imagen;
    }

    public String getModelo() {
        return modelo;
    }

    public void setModelo(String modelo) {
        this.modelo = modelo;
    }

    public String getMarca() {
        return marca;
    }

    public void setMarca(String marca) {
        this.marca = marca;
    }

    public int getCv() {
        return cv;
    }

    public void setCv(int cv) {
        this.cv = cv;
    }

    public int getKm() {
        return km;
    }

    public void setKm(int km) {
        this.km = km;
    }

    @Override
    public String toString() {
        return getModelo();
    }
}

File: Kotlin/T1/codigos2223/Clases_RET/src/main/kotlin/model/Coche.kt
package model

// final
// open
class Coche(var marca: String, var modelo: String) {

    // atributos
    var bastidor: String? = null;
    private var cv: Int = 0
        get() = field
        set(cv) {
            println("Valor modificado")
            field = cv
        }
    var cc: Int = 0;
    lateinit var propietario: Propietario;

    // constructores
    // public Coche(){}
    /*constructor(marca: String, modelo: String){
        this.marca = marca;
        this.modelo = modelo;
    }*/

    constructor(marca: String, modelo: String, bastidor: String) : this(marca, modelo) {
        this.bastidor = bastidor
    }

    // bloque ejecutado tras cualquier constructor
    /*init {
        propietario = Propietario("Borja","Martin","1234A")
    }*/
    // funciones

    var calcularVelocidad: ((Int)->Unit)? = null;

    fun aumentarVelocidad(velocidad: Int ){
        // calcularVelocidad(6) --> no se como se comporta Int -> Unit
        calcularVelocidad?.invoke(this.cc)
    }

    fun asignarPropietario(propietario: Propietario){
        this.propietario = propietario;
    }

    fun calcularParMotor (calculoPar:(Int)->Int, aceleracion: Int){
        calculoPar(aceleracion);
    }

    var mostrarDatos: () -> Unit = {
        println("Marca: $marca")
        println ("Modelo: ${modelo}")
        println ("CV: ${cv}")
        println ("CC: ${cc}")
    }

    // getter setter

    fun setCV(cv: Int) {
        this.cv = cv;
    }

    fun getCV(): Int {
        return this.cv
    }

    // toString
    override fun toString(): String {
        return super.toString()
    }
}